### DISCUSSION TITLE
Is the "Design and Roadmap" document still accurate?

### DISCUSSION THREAD

**Discussion by analogduck** [EFForg/privacybadger#2925] [Q&A]
Back in October of 2020, [it was announced](https://www.eff.org/deeplinks/2020/10/privacy-badger-changing-protect-you-better) that local learning was being disabled. The [PRIVACY BADGER DESIGN AND ROADMAP](https://github.com/EFForg/privacybadger/blob/master/doc/DESIGN-AND-ROADMAP.md) document still talks about tracking, counting, and blocking third-party trackers. Does that document need to be updated, or is that still the architecture? My understanding of that announcement about local learning is that Privacy Badger is not longer doing that counting and blocking. Is that correct?

**Comment by ghostwords** [comment]
Hello! The design document is still accurate. Privacy Badger no longer learns locally **by default**. As we wrote in the ["Privacy Badger is changing" blog post](https://www.eff.org/deeplinks/2020/10/privacy-badger-changing-protect-you-better), you can still choose to opt in to local learning and have the exact same Badger experience as before.

Privacy Badger still automatically learns to block invisible trackers. The default mode of learning is "remote" learning, powered by [Badger Sett](https://www.eff.org/deeplinks/2018/08/giving-privacy-badger-jump-start) and [Badger Swarm](https://www.eff.org/deeplinks/2023/10/privacy-badger-learns-block-ever-more-trackers). In addition, local learning can be enabled with a click on the options page.

You may also want to visit [our FAQ](https://privacybadger.org/#faq). Let me know if you have any questions.

**Reply by analogduck** [reply]
Thanks, that helps. I think my confusion was around what "learning" means or, more specifically, when and where the learning is happening. So my understanding is:

Before August 2018, all learning was performed inside of the Privacy Browser extension installed locally in my browser. The extension started out with no knowledge about what domains to block.

In August 2018, Badger Sett was released. This utility is used by the Privacy Badger development team. It installs the Privacy Badger extension into Selenium, hits a list of popular websites, and that generates a block list. This is run before each release of the Privacy Badger extension, and the resulting seed list is embedded in the extension that users install in their browsers. The only way for users to update that seed list is to install a newer release of the Privacy Badger extension. At this point in the history, Privacy Badger was still detecting new trackers locally in users' browsers and building a more extensive block list that was specific to that local browser.

In October 2020, local learning was disabled by default. It can still be enabled but, for most users, the only trackers that Privacy Badger is blocking are the trackers defined in the `seed.json` file in the Privacy Badger release that they had installed.

In October 2023, the Privacy Badger team started using Badger Swarm. This allowed for the generation of a larger seed list, but does not change the way that Privacy Badger runs in a given user's browser.

To summarize, the learning went from "only in the user's browser" to "pre-release by the dev team and also in the user's browser" to "only pre-release by the dev team (with the option to also learn locally)". The Privacy Badger extension has always been responsible for performing the learning, but when and where that learning happens is what's changed.

Is that accurate?

Thanks, and thanks for all of the work you've put into this.

**Reply by ghostwords** [reply]
There is a good summary of Privacy Badger's evolution in the [Badger Swarm announcement blog post](https://www.eff.org/deeplinks/2023/10/privacy-badger-learns-block-ever-more-trackers).

**Reply by ghostwords** [reply]
>the only trackers that Privacy Badger is blocking are the trackers defined in the seed.json file in the Privacy Badger release that they had installed

Users get their Privacy Badger's tracker knowledge regularly refreshed (at this time as part of Privacy Badger updates).

**Reply by ghostwords** [reply]
>The Privacy Badger extension has always been responsible for performing the learning, but when and where that learning happens is what's changed.

Yep!

You're welcome! And thank YOU for your interest and support!
